crypto/rand/rand_unix.c

```diff
@@ -273,6 +283,111 @@ int syscall_random(void *buf, size_t buflen)
     return -1;
 }
 
+#if !defined(OPENSSL_RAND_SEED_NONE) && defined(OPENSSL_RAND_SEED_DEVRANDOM)
+static const char *devrandom_paths[] = { DEVRANDOM };
+static struct {
+    int fd;
+    dev_t dev;
+    ino_t ino;
+    mode_t mode;
+    dev_t rdev;
+} devrandom_files[OSSL_NELEM(devrandom_paths)];
+static int devrandom_keep_open = 1;
+
+/*
+ * Verify that the file descriptor associated with the random source is
+ * still valid. The rationale for doing this is the fact that it is not
+ * uncommon for daemons to close all open file handles when daemonizing. So
+ * the handle might have been closed or even reused for opening another file.
+ */
+static int check_devrandom_device(size_t n)
+{
+    struct stat st;
+    const int fd = devrandom_files[n].fd;
+
+    return fd != -1
+           && fstat(fd, &st) != -1
+           && devrandom_files[n].dev == st.st_dev
+           && devrandom_files[n].ino == st.st_ino
+           && devrandom_files[n].mode == st.st_mode
+           && devrandom_files[n].rdev == st.st_rdev;
+}
+
+/*
+ * Open a random device if requried and return the FD or -1 on error
+ */
+static int get_devrandom_device(size_t n)
+{
+    if (!check_devrandom_device(n)) {
+        struct stat st;
+        const int fd = open(devrandom_paths[n], O_RDONLY);
+
+        if (fd != -1 && fstat(fd, &st) != -1) {
+            devrandom_files[n].fd = fd;
+            devrandom_files[n].dev = st.st_dev;
+            devrandom_files[n].ino = st.st_ino;
+            devrandom_files[n].mode = st.st_mode;
+            devrandom_files[n].rdev = st.st_rdev;
+        } else {
+            devrandom_files[n].fd = -1;
```

Comment:
Formally speaking there is possibility for resource leak here. In case fd != -1 while return value from fstat(fd) is.

Reply:
Good catch!

Reply:
Now I see why you were saying 'Formally speaking'. Looking at the error codes listed in stat(2) there really seems to be only a hypothetical chance that fstat() might fail in this situation. :-)

Reply:
ENOMEM would be possible.  I'll fix.

Reply:
It doesn't matter what is error condition or how impossible it is. Mere fact of checking the return value in given sequence **formally** makes it a possible resource leak.